daemons/attrd/attrd_utils.c

```diff
@@ -375,10 +381,43 @@ free_action(gpointer data)
 {
     struct confirmation_action *action = (struct confirmation_action *) data;
     g_list_free_full(action->respondants, free);
+    mainloop_timer_del(action->timer);
     free(action);
     action = NULL;
 }
 
+/* Remove the client from the expected_confirmations table if it times out before
+ * receiving all confirmations.  We set the timeout to 6000ms, which is 1000ms
+ * longer than the timeout used in send_attrd_request/connect_and_send_attrd_request
+ * (used by attrd_updater) just to give everything a moment to settle.
+ *
+ * The client will be removed from the waitlist by attrd_remove_client_from_waitlist
+ * which gets called by attrd_ipc_close when the client disconnects.
+ */
+static gboolean
+confirmation_timeout_cb(gpointer data)
+{
+    struct confirmation_action *action = (struct confirmation_action *) data;
+
+    GHashTableIter iter;
+    gpointer value;
+
+    if (expected_confirmations == NULL) {
+        return G_SOURCE_REMOVE;
+    }
+
+    g_hash_table_iter_init(&iter, expected_confirmations);
+
+    while (g_hash_table_iter_next(&iter, NULL, &value)) {
+        if (value == action) {
+            g_hash_table_iter_remove(&iter);
+            break;
+        }
+    }
```

Comment:
We still have to reply to the client, just with a timeout failure

Reply:
Fixed, but it doesn't look like the time out status code ever makes it back up to the command line tool. It looks like as long as we were able to get some reply from the server, that's good enough and that's the error code that gets used. Whatever actual status code is hiding in the XML reply doesn't matter.